lib/Partitioner/Partitioner.cpp

```diff
@@ -485,14 +520,124 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+DeviceIDTy Partitioner::assignLogicalDeviceID(NodeToFunctionMap &mapping) {
+  DeviceIDTy logicalDeviceID = 0;
+
+  std::map<BackendKind, std::vector<Function *>> backendFuncMap;
+  for (auto &func : mapping.getPartitions()) {
+    // Traverse the paritions, and get list of partitions with each backendKind.
+    auto backendKind = mapping.getPartitionBackendKind(func);
+    if (backendFuncMap.find(backendKind) == backendFuncMap.end()) {
+      backendFuncMap.emplace(backendKind, std::vector<Function *>{func});
+    } else {
+      backendFuncMap[backendKind].push_back(func);
+    }
+  }
+
+  // For each type of the backend, assign the logicalDevice ID.
+  for (const auto &p : backendFuncMap) {
+    if (mapping.getPartitions().size() <= backendMap_[p.first].num) {
+      // There is enough device with this backendKind, no need to adjust the
+      // logical ID.
+      for (auto &func : p.second) {
+        mapping.appendLogicalDeviceID(func, logicalDeviceID++);
+      }
+      continue;
+    }
+    // Get the list of functions with current BackendKind, and sort it based on
+    // used memory from min to max.
+    std::vector<std::pair<Function *, uint64_t>> nodeSize;
+    for (size_t i = 0, e = p.second.size(); i < e; i++) {
+      Function *function = p.second[i];
+      uint64_t totalMem = mapping.getGraphMemInfo(function).getTotalMemSize();
+      nodeSize.push_back(std::make_pair(p.second[i], totalMem));
+    }
+    std::sort(nodeSize.begin(), nodeSize.end(), sortMinMemory);
+
+    // Assume we have n devices(NOTE: here the n devices have the same avaiable
+    // memory size, and the following algorithm can find the accurate result. If
+    // the momery size are differnt, this assignment issue will be a NP problem
+    // -- multiple knapsack problem, and the following algorithm becomes greedy
+    // and the result may not be optimal), and m partitions, where m > n. If
+    // these m partitions can be assigned to n devices, there must be 1 device
+    // have at least (m - 1)/n + 1 partitions(Pigeonhole principle). Based on
+    // this theory, the algorithm is: Given N devices, and M partitions:
+    // Step 1 : sort the partitions from min to max based on their memory usage.
+    std::sort(nodeSize.begin(), nodeSize.end(), sortMinMemory);
+    // Step 2 : let n = N, m = M.
+    size_t m = p.second.size();
+    size_t n = backendMap_[p.first].num;
+    while (m > 0) {
+      // Step 3 : find the first k partitions whose total memory usage still
+      // under the memory limitation (k should be max).
+      uint64_t usedMem = 0;
+      size_t numOfPartitionsWithSameID = (m - 1) / n + 1;
+      size_t start = p.second.size() - m;
+      size_t i;
+      for (i = start; i < p.second.size(); i++) {
+        if (usedMem + nodeSize[i].second > backendMap_[p.first].memSize) {
+          break;
+        }
+        usedMem += nodeSize[i].second;
+      }
+      // Step 4 : if k found in step 3 is larger than (m - 1) / n + 1, this
```

Comment:
larger -> smaller,
Also it might be helpful to call out in the comment that k = i- start.